internal/registry/registry_cache.go

```diff
@@ -204,6 +208,125 @@ func (r *Cache) validateCache() error {
 	return errors.New("invalid registry cache remote")
 }
 
+func (r *Cache) CreateCommit(b Buildpack, username, msg string) error {
+	r.logger.Debugf("Creating commit in registry cache")
+
+	if msg == "" {
+		return errors.New("invalid commit message")
+	}
+
+	repository, err := git.PlainOpen(r.Root)
+	if err != nil {
+		return errors.Wrap(err, "opening registry cache")
+	}
+
+	w, err := repository.Worktree()
+	if err != nil {
+		return errors.Wrapf(err, "reading (%s)", r.Root)
+	}
+
+	index, err := r.writeEntry(b)
+	if err != nil {
+		return errors.Wrapf(err, "writing (%s)", index)
+	}
+
+	if index == "" {
+		return nil
+	}
+
+	if _, err := w.Add(index); err != nil {
+		return errors.Wrapf(err, "adding (%s)", index)
+	}
+
+	if _, err := w.Commit(msg, &git.CommitOptions{
+		Author: &object.Signature{
+			Name:  username,
+			Email: "",
+			When:  time.Now(),
+		},
+	}); err != nil {
+		return errors.Wrapf(err, "committing")
+	}
+
+	return nil
+}
+
+func (r *Cache) writeEntry(b Buildpack) (string, error) {
+	var ns string = b.Namespace
+	var name string = b.Name
+
+	var indexDir string
+	switch {
+	case len(name) == 0:
+		return "", errors.New("empty buildpack name")
+	case len(name) == 1:
+		indexDir = "1"
+	case len(name) == 2:
+		indexDir = "2"
+	case len(name) == 3:
+		indexDir = "3"
+	default:
+		indexDir = filepath.Join(name[:2], name[2:4])
+	}
+
+	if len(ns) == 0 {
+		return "", errors.New("empty buildpack namespace")
+	}
+
+	index := filepath.Join(r.Root, indexDir, fmt.Sprintf("%s_%s", ns, name))
+
+	if _, err := os.Stat(filepath.Join(r.Root, indexDir)); os.IsNotExist(err) {
+		if err := os.MkdirAll(filepath.Join(r.Root, indexDir), 0755); err != nil {
+			return "", errors.Wrapf(err, "creating directory structure for: %s/%s", ns, name)
+		}
+	} else {
+		if _, err := os.Stat(index); err == nil {
+			entry, err := r.readEntry(ns, name)
+			if err != nil {
+				return "", errors.Wrapf(err, "reading existing buildpack entries")
+			}
+
+			availableBuildpacks := entry.Buildpacks
+
+			if len(availableBuildpacks) != 0 {
+				if availableBuildpacks[len(availableBuildpacks)-1].Version == b.Version {
+					return "", errors.Wrapf(err, "same version exists, upgrade the version to add")
+				}
+			}
+		}
+	}
```

Comment:
A lot of this seems to be duplicated from the `readEntry()` func. If we could pull out some or all of this into funcs that are used in both `readEntry` and `writeEntry`, or if we could call `readEntry()` within `writeEntry` to simplify a lot of this, that would be great

Reply:
Yeah actually we can separate out some parts of readEntry() and reuse it here. Would definitely a lot cleaner.